contracts/src/token/erc20/utils/safe_erc20.rs

```diff
@@ -0,0 +1,229 @@
+        function allowance(address owner, address spender) external view returns (uint256);
+    }
+}
+
+sol_storage! {
+    /// State of the SafeErc20 Contract.
+    pub struct SafeErc20 {}
+}
+
+/// NOTE: Implementation of [`TopLevelStorage`] to be able use `&mut self` when
+/// calling other contracts and not `&mut (impl TopLevelStorage +
+/// BorrowMut<Self>)`. Should be fixed in the future by the Stylus team.
+unsafe impl TopLevelStorage for SafeErc20 {}
+
+#[public]
+impl SafeErc20 {
+    /// Transfer `value` amount of `token` from the calling contract to `to`. If
+    /// `token` returns no value, non-reverting calls are assumed to be
+    /// successful.
+    pub fn safe_transfer(
+        &mut self,
+        token: Address,
+        to: Address,
+        value: U256,
+    ) -> Result<(), Error> {
+        let encoded_args = (to, value).abi_encode_params();
+        let selector = function_selector!("transfer", Address, U256);
+        let data = [&selector[..4], &encoded_args].concat();
+
+        self._call_optional_return(token, &data)
+    }
+
+    /// Transfer `value` amount of `token` from `from` to `to`, spending the
+    /// approval given by `from` to the calling contract. If `token` returns
+    /// no value, non-reverting calls are assumed to be successful.
+    pub fn safe_transfer_from(
+        &mut self,
+        token: Address,
+        from: Address,
+        to: Address,
+        value: U256,
+    ) -> Result<(), Error> {
+        let encoded_args = (from, to, value).abi_encode_params();
+        let selector =
+            function_selector!("transferFrom", Address, Address, U256);
+        let data = [&selector[..4], &encoded_args].concat();
+
+        self._call_optional_return(token, &data)
+    }
+
+    /// Increase the calling contract's allowance toward `spender` by `value`.
+    /// If `token` returns no value, non-reverting calls are assumed to be
+    /// successful.
+    pub fn safe_increase_allowance(
+        &mut self,
+        token: Address,
+        spender: Address,
+        value: U256,
+    ) -> Result<(), Error> {
+        let erc20 = IERC20::new(token);
+        let call = Call::new_in(self);
+        let old_allowance = erc20.allowance(call, address(), spender).or(
+            Err(Error::SafeErc20FailedOperation(SafeErc20FailedOperation {
+                token,
+            })),
+        )?;
+        self.force_approve(token, spender, old_allowance + value)
+    }
+
+    /// Decrease the calling contract's allowance toward `spender` by
+    /// `requested_decrease`. If `token` returns no value, non-reverting
+    /// calls are assumed to be successful.
+    pub fn safe_decrease_allowance(
+        &mut self,
+        token: Address,
+        spender: Address,
+        requested_decrease: U256,
+    ) -> Result<(), Error> {
+        let erc20 = IERC20::new(token);
+        let call = Call::new_in(self);
+        let current_allowance =
+            erc20.allowance(call, address(), spender).or({
```

Comment:
Refactor `erc20.allowance` call into separate fn for both `safe_decrease_allowance` and `safe_increase_allowance`.